### ISSUE TITLE
Java 7 NIO filewatcher

### ISSUE THREAD

**Issue by fommil** [ensime/ensime-server#1217]
@aemoncannon creating this just because one of us will eventually work on this.

We should put this in a separate repo so it can build under java 7. I'll try to get a docker image built for that.

One of the upstream tickets in #1042 points to some prior art in this area. Also, I think sbt might have some Java 7 file watchers, so we could potentially depend on their IO library if its in there.

**Comment by fommil** [comment]
specifically https://issues.apache.org/jira/browse/VFS-537?focusedCommentId=14491540&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14491540

**Comment by jvican** [comment]
I will pick this up and try to solve this issue. 
  
After some discussion at gittter, we have come up with two possible paths to solve this:
  
1. Detect Java 1.7+ automatically at runtime and then use the `FileWatcher` API. If only Java 1.6 is available, use sync monitoring of files/folders as it is now on `ensime-server`.
2. Since there are a lot of bugs related to VFS, it could be better to just roll out our own implementation of JAR/ZIP scanning and files/folders monitoring. Even better, I could contribute to the `rapture` project to make this functionality available there and then use that as a dependency. I am currently discussing this with @propensive and I don't know if it's feasible or if it's not.
  
Nevertheless, I think that the second option should be the way to go to avoid more problems with VFS in the future and not to depend on a project that seems unmantained.

**Comment by jvican** [comment]
Already working on the second option after a fruitful conversation with @propensive. Hope to have this ASAP. I will start by implementing the ZIP scanner (https://github.com/propensive/rapture/issues/52) and the move on to the JAR one and the monitoring API.

**Comment by fommil** [comment]
@jvican given that the filewatchers are reasonably high priority and I need something similar in https://github.com/fommil/sbt-big-project can we say that the rapture stuff is more likely a strategic replacement for Apache VFS and shouldn't block any work on async file scanners?

**Comment by fommil** [comment]
@jvican don't forget about random access within zip/jar files and the ability to have a URL-like naming convention (which goes into the DB, so would be ideally a like-for-like replacement for the current VFS URLs)

**Comment by jvican** [comment]
Hi @fommil, we can proceed as you prefer. I am already working on the rapture thing and I could have it in one week and a half tops (everything). Perhaps, even sooner. I'm taking into account the random access within zip/jar files and the URL-like naming convention but I will discuss this with @propensive. I imagine he will prefer a more generic API that may not suit us 100%. However, if necessary I will code on top of the rapture module and fit exactly the conventions/needs of `ensime-server`.

**Comment by fommil** [comment]
awesome! But you're not working on the filewatchers right now though, right?

They will require a standalone project that builds on Java 7, so they won't fit into rapture unless they are an extension module.

**Comment by jvican** [comment]
No, I was planning to work before on the JAR/ZIP scanner, but if you really need them I will hurry and give them priority. Hope to have them before :+1: 

I will create that standalone project then, are you planning to move to JDK7 any time soon, right? Then, we could merge the projects or just declare mine as a dependency.

**Comment by fommil** [comment]
The other way around, the jar/zip scanner is not a priority so take your time and do it right :smile:

I have a priority on the JDK7 scanners for other reasons, so it is more likely that our efforts will overlap. i.e. focus on the jar/zip stuff

**Comment by jvican** [comment]
Yep, sorry, I mean the filewatchers. I will give them priority if you really need them. How much time can you wait for them?
  
I could just focus on that before the JAR/ZIP scanner. Therefore, I can help you.

**Comment by fommil** [comment]
@jvican I need them yesterday :smile: I'll try to set up an empty repo tonight for them. They need to be Apache 2.0 licence because I need them in `sbt-big-project`.

It's fine if you focus on the jar/zip stuff... I will probably work on them.

**Comment by propensive** [comment]
The file watchers will be an extension module. That's how things are usually implemented in Rapture. There's a single, big Rapture project with everything in, but also a hierarchy of smaller projects, so it would be straightforward to depend on just the filesystem stuff, and not pull in a vast number of dependencies.

**Comment by fommil** [comment]
@propensive would you be interested in looking through the ensime server code to see where we could use other parts of rapture instead of hand-rolled or guava/commons?

**Comment by fommil** [comment]
@aemoncannon @propensive @jvican we now have the docker image `ensime/ensime:v2.0` which contains OpenJDK 7 and Emacs 24.{4, 5}. We won't use this for `ensime-server` but we can use it for dependency projects. I'll probably use it in my `sbt-big-project` build, where I'll be writing NIO hacky FileWatchers, but I imagine that we'll eventually get something tidier in Rapture by the sounds of it.

**Comment by fommil** [comment]
sorry that's `ensime/ensime:v2.x`

**Comment by fommil** [comment]
empty repo set up at https://github.com/ensime/java7-file-watcher and @lxol has a prototype, while @jvican works on the abstract file system layer in Rapture.

**Comment by fommil** [comment]
@lxol note that it is a scala project. If it turns out that you only have java files, then just turn off the scala stuff by doing

```
crossPaths := false
autoScalaLibrary := false
```

**Comment by fommil** [comment]
@aemoncannon FYI @lxol is was great progress on this (with @matarrese) at the hack day on Saturday. The tests in https://github.com/ensime/java7-file-watcher/pull/2 are almost passing.

**Comment by fommil** [comment]
@lxol I've got a dev branch https://github.com/fommil/ensime-server/tree/java7-file-watcher pulling this into `ensime-server` but two problems remain before we can merge:

1. we seem to be using one thread per watcher !!!! This isn't going to work, how can we share one thread for all listeners?
2. I'm getting a failure in the `MissingJarTargetTest` integration test, which would imply that the unit test `"be able to start up from a non-existent base file"` isn't doing everything it should be.

**Comment by lxol** [comment]
@fommil
 With threads you are right it is one one watcher per thread but it can
monitor multiple recursive directories  and notify multiple listeners. But
I think we can move thread management to ensime-server from the watcher if
that's not acceptable.
I will have a look at that and the failed test  tomorrow, didn't really
have time today.

**Comment by fommil** [comment]
@lxol awesome! Even with that many threads (100 for ensime-server) it still felt much faster (and used less CPU) than the java 6 threads!

**Comment by lxol** [comment]
ok , I had a  quick look the the thread issue.
When I developed Java7 watcher I tried  to mimic the Apache's
DefaulFileMonitor that does create one  new thread per watcher. What I
didn't check is that NIO WatchService use a background Poller thread per
watcher so we end up with twice the number of threads.
For example ensime-server project has around 70 DefaultFileMonitor threads
if start with Java6 and more than a hundred with Java7.

I can remove an extra thread per watcher but it won't be less that we have
already with DefaultFileMonitor.

**Comment by fommil** [comment]
we can redesign the API for the Java7 usage. The reason we have one watcher per directory / jar is to workaround VFS bugs (we need to watch the root incase it gets deleted), but I don't think we need to do that anymore for Java 7 do we?

I guess there are two steps here:

- thread usage parity (we can merge at this point)
- reduce the thread count to, ideally, one

**Comment by lxol** [comment]
ok.

I am not aware of that bug in Java 7 watcher and I had successful multi
directory per watcher test.

**Comment by fommil** [comment]
if you watch `/a/b` and it is deleted (even if `/a` is deleted also) is it ok? That's really important because the target directories get deleted by sbt.

**Comment by lxol** [comment]
I am not quite sure now. I found that Java7 file watcher reports a jar file
in a completely  unrelated directory in your failed test. I will
investigate it further.

**Comment by lxol** [comment]
the watcher reported a file name relative to the watched directory instead
of the absolute path. I submitted a PR to fix that.

@fommil  I guess the case you described covered by "survive deletion of the
watched directory" test from FileWatcherSpec and java7 watcher passed it.

I will go about the  thread issue now.

**Comment by lxol** [comment]
It looks like there is no way around of 2 threads per watcher. We create an
instance of WatchService that uses an internal background poll thread. Then
we use a blocking `take` function in a loop to process WatchEvents.

https://github.com/ensime/java7-file-watcher/blob/master/src/main/java/org/ensime/filewatcher/FileMonitor.java#L231

On the plus side we can register directories recursively with a new watcher.

**Comment by fommil** [comment]
is there no async way to get the information from the watcher?

**Comment by fommil** [comment]
what if we have a single thread that polls all our watchers in a loop?

**Comment by lxol** [comment]
I think we can use a non-blocking `poll` for that. But why do we need to consume extra cpu cycles if we can do the same with blocked background threads?

**Comment by fommil** [comment]
imagine you have 1,000+ directories to watch (like me) :smile:

**Comment by lxol** [comment]
You'll make a cpu really busy if you poll 1,000+ watchers every second then... 

How difficult would it be to utilize a recursive watcher in ensime-server?

**Comment by lxol** [comment]
Actually I think we can redesign a watcher to register all possible directories and selectors in one watcher and maintain a list of listeners with associated watched directories and selectors. In that case we can notify  only listeners that have matching dirs and selectors for the WatchEvent.

**Comment by lxol** [comment]
or use several watchers that  specialize  on particular selectors. "jar", ".scala" watchers, etc...

**Comment by fommil** [comment]
that would be nice, but are you sure it handles the case where the base is deleted and recreated?

**Comment by lxol** [comment]
The watcher detects the deletion of the base directory. We can react by
recreating  and re-regestering  the base, the same technique that we use
now.

**Comment by fommil** [comment]
OK, that sounds good! Code it as if it was going to be a complete replacement for what we currently have and we can do the swap at the right level.

**Comment by lxol** [comment]
ok

**Comment by lxol** [comment]
I've rewritten the watcher in scala. Still wip though, 5 out of 16 tests fail. I am dealing now with some edge cases when base is a file. It is tricky because Java7 WatchService can register only directories.

https://github.com/lxol/java7-file-watcher/commit/3e04acbd2ae53dcb64c7a1eedb911d672d7cd5a6

**Comment by fommil** [comment]
great stuff! Looking forward to this :smile:

**Comment by lxol** [comment]
@fommil  can we add an issue tracker to ensime/java7-file-watcher. I noticed a difference in `inotify` behaviour on OS X and also wanted to investigate if we affected by WatchService bug on Linux  http://blog.omega-prime.co.uk/?p=161. It would be nice to create separate issues for that.

**Comment by fommil** [comment]
I'm out this week, definitely ok to use ensime-server the track this stuff.

**Comment by lxol** [comment]
I started to use a new watcher with ensime-server. https://github.com/lxol/ensime-server/commit/0847ba401526bb74043672e70e475cd215da5108
The first thing I noticed in emacs is that the watcher reacts to `.#`  temporary files. Is that a good idea to be more specific with the selections and use regexp instead of just extensions?

**Comment by fommil** [comment]
Good idea

**Comment by lxol** [comment]
another observation - with a new watcher, the ensime-server project has only 4 watcher related threads. 
~ 70 with current apache DefaultFileMonitor

**Comment by rorygraves** [comment]
Awesome observations!

**Comment by rorygraves** [comment]
Sounds like this is going to be a huge win.

**Comment by fommil** [comment]
:-)

**Comment by jvican** [comment]
I'd love to have this! Keep up the good work and hope to have a release soon! :+1:

**Comment by fommil** [comment]
Do we have an OS X Travis test set up for the file watchers?

**Comment by lxol** [comment]
It would be nice to have OS X Travis tests. Java7 WatchService is platform dependent and we've already seen the differences between Win, Linux and OS X.

**Comment by andrebeat** [comment]
A bit off-topic but I've noticed that you've created another project to depend on Java 7 APIs. I have a project where I've setup conditional compilation based on JVM version ([build.sbt]) and also conditional selection of appropriate classes at runtime ([Adder.scala]). Just thought this could be useful.

[build.sbt]: https://github.com/andrebeat/scala-pool/blob/master/build.sbt#L30
[Adder.scala]: https://github.com/andrebeat/scala-pool/blob/master/src/main/scala/io/github/andrebeat/pool/Adder.scala

**Comment by fommil** [comment]
It won't work for us because our docker images don't have Java 7 and it's a huge hassle to change that right now. Thanks anyway.

**Comment by fommil** [comment]
@aemoncannon is this working for you at work?
